src/ai/backend/storage/services/artifacts/huggingface.py

```diff
@@ -118,6 +120,149 @@ def log_progress(offset: int, final: bool = False) -> None:
+
+                        # Fill total size from response headers if still unknown
+                        if total is None:
+                            if resp.status == 200:
+                                content_length = resp.headers.get("Content-Length")
+                                if content_length and content_length.isdigit():
+                                    total = int(content_length)
+                            elif resp.status == 206:
+                                content_range = resp.headers.get(
+                                    "Content-Range"
+                                )  # e.g. "bytes 123-456/789"
+                                if content_range and "/" in content_range:
+                                    try:
+                                        total = int(content_range.split("/")[-1])
+                                    except ValueError:
+                                        pass
+                            # TODO: Handle else case
+
+                        async for chunk in resp.content.iter_chunked(self._chunk_size):
+                            if not chunk:
+                                continue
+                            offset += len(chunk)
+                            progress_logger(offset)
+                            yield chunk
+
+                    # total unknown
+                    if total is None:
+                        progress_logger(offset, final=True)
+                        log.warning("Skipped download of %s since total size is unknown", self._url)
+                        break
+
+                    # Completed
+                    if offset >= total:
+                        progress_logger(offset, final=True)
+                        break
+
+                    # Unexpected early EOF → retry
+                    raise aiohttp.ClientPayloadError("Early EOF before Content-Length")
+
+                except _DOWNLOAD_RETRIABLE_ERROR as e:
+                    retries += 1
+                    if retries > self._max_retries:
+                        progress_logger(offset, final=True)
+                        raise aiohttp.ClientPayloadError(
+                            f"Exceeded retries while downloading {self._url} at offset={offset}"
+                        ) from e
+
+                    log.warning(
+                        "Download retry {}/{} at offset {:.1f} MiB (backoff={:.1f}s, err={})",
+                        retries,
+                        self._max_retries,
+                        offset / _MiB,
+                        backoff,
+                        e.__class__.__name__,
+                    )
+                    await asyncio.sleep(backoff)
+                    backoff = min(backoff * 2, 30.0)
+
+                    # Refresh metadata before retry
+                    await _probe_head()
+                    continue
+        finally:
+            await session.close()
```

Comment:
A very large method has appeared, so please break down the private method well according to role.